### ISSUE TITLE
proxy problem

### ISSUE THREAD

**Issue by timsgardner** [arcadia-unity/Arcadia#163]
Looks like another problem with proxy. The following (encountered during port of Riddley) does not work:

```clojure
(proxy [clojure.lang.CljCompiler.Ast.ObjMethod]
    [(clojure.lang.CljCompiler.Ast.ObjExpr. nil) nil])
```
throws:
```
System.TypeLoadException: Could not load type 'user.proxy$ObjMethod$FF19274A$_INTERP' from assembly 'proxy22, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
  at clojure.lang.Compiler.eval (System.Object form) [0x00000] in <filename unknown>:0 
  at clojure.lang.Compiler.eval (System.Object form) [0x00000] in <filename unknown>:0 
  at clojure/core$eval__3013.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/repl$eval_to_string$fn__27866__27870.invoke () [0x00000] in <filename unknown>:0 
  at arcadia/repl$eval_to_string__27873.invoke (System.Object ) [0x00000] in <filename unknown>:0 
  at arcadia/repl$repl_eval_print$fn__27882__27886.invoke () [0x00000] in <filename unknown>:0 
```
Note that
```clojure
(clojure.lang.CljCompiler.Ast.ObjMethod.
  (clojure.lang.CljCompiler.Ast.ObjExpr. nil) nil)
```
totally works.

On JVM,
```clojure
(import '[clojure.lang Compiler Compiler$ObjMethod Compiler$ObjExpr])
(proxy [Compiler$ObjMethod] [(Compiler$ObjExpr. nil) nil])
```
works.

**Comment by timsgardner** [comment]
How would we rate the difficulty of solving this bug, on a scale from 0 - 10?

**Comment by nasser** [comment]
Narrowing it down: `ObjMethod` is an abstract class and that could be a factor? You shouldn't even be able to make instances of abstract classes, so I am surprised that 

```clojure
(clojure.lang.CljCompiler.Ast.ObjMethod.
  (clojure.lang.CljCompiler.Ast.ObjExpr. nil) nil)
```
 works.

Repeating the experiment with `FnMethod` and `NewInstanceMethod`, concrete subclasses of `ObjMethod`:

```clojure
(proxy [clojure.lang.CljCompiler.Ast.FnMethod]
  [nil nil])
;; #<proxy$FnMethod$FF19274A$_INTERP user.proxy$FnMethod$FF19274A$_INTERP>

(proxy [clojure.lang.CljCompiler.Ast.NewInstanceMethod]
  [nil nil])
;; clojure.lang.Compiler+CompilerException: System.NotSupportedException: Operation is not supported.
;;   at System.Reflection.Emit.TypeBuilder.GetConstructors (BindingFlags bindingAttr) [0x00023] in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection.Emit/TypeBuilder.cs:883 
;;   at clojure.lang.Reflector.GetConstructors (System.Type targetType, Int32 arity) [0x00000] in <filename unknown>:0 
;;   at clojure.lang.Reflector.GetMatchingConstructor (IPersistentMap spanMap, System.Type targetType, IList`1 args, System.Int32& ctorCount) [0x00000] in <filename unknown>:0 
;;   at clojure.lang.CljCompiler.Ast.NewExpr.ComputeCtor () [0x00000] in <filename unknown>:0 
;;   at clojure.lang.CljCompiler.Ast.NewExpr..ctor (System.Type type, System.Collections.Generic.List`1 args, IPersistentMap spanMap) [0x00000] in <filename unknown>:0 
;;   at clojure.lang.CljCompiler.Ast.NewExpr+Parser.Parse (clojure.lang.CljCompiler.Ast.ParserContext pcon, System.Object frm) [0x00000] in <filename unknown>:0 
;;   at clojure.lang.Compiler.AnalyzeSeq (clojure.lang.CljCompiler.Ast.ParserContext pcon, ISeq form, System.String name) [0x00000] in <filename unknown>:0 , compiling: (NO_SOURCE_PATH:0:0)
```

Very strange, unexpected behavior. I would put this at an 8/10 difficulty. It is going to involve bytecode spelunking.

**Comment by timsgardner** [comment]
Ok fuck it then. I'll just do whatever needs doing for this port in C#, we can add proxy to the list of Known Antlion Pits 
![image](https://cloud.githubusercontent.com/assets/451342/18943470/cbae1786-85ed-11e6-8bae-0dbe3d7eecf2.png)

**Comment by nasser** [comment]
There are so many Known Antlion Pits, and so many more unknown...

**Comment by nasser** [comment]
TLDR behavior seems more consistent and errors are more meaningful. I don't think these are issues anymore.

On the current `beta`

```clj
(proxy [clojure.lang.CljCompiler.Ast.ObjMethod]
    [(clojure.lang.CljCompiler.Ast.ObjExpr. nil) nil])
```

Now throws

```
NotSupportedException: Parent does not have a default constructor. The default constructor must be explicitly defined.
```

Which is more meaningful. As noted above `ObjMethod` is an abstract class and has special rules about it.

Moreover

```clj
(clojure.lang.CljCompiler.Ast.ObjMethod.
  (clojure.lang.CljCompiler.Ast.ObjExpr. nil) nil)
```

now throws

```
System.InvalidOperationException: No constructor in type: ObjMethod with 2 arguments
```


Repeating the other experiments

```clj
(proxy [clojure.lang.CljCompiler.Ast.FnMethod]
  [nil nil])
;; => #object[proxy$FnMethod$FF19274A$_INTERP 0xb6733d00 "user.proxy$FnMethod$FF19274A$_INTERP"]

(proxy [clojure.lang.CljCompiler.Ast.NewInstanceMethod]
  [nil nil])
;; Could not load type 'user.proxy$NewInstanceMethod$FF19274A$_INTERP' from assembly 'proxy15, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' because the parent type is sealed.
```

`NewInstanceMethod` is in fact sealed, so I don't know that these are actually bugs?
